sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala

```diff
@@ -314,4 +314,24 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
       }
     }
   }
+
+  test("SPARK-11103: Filter applied on merged Parquet schema with new column fails") {
+    import testImplicits._
+
+    withSQLConf(SQLConf.PARQUET_FILTER_PUSHDOWN_ENABLED.key -> "true",
+      SQLConf.PARQUET_SCHEMA_MERGING_ENABLED.key -> "true") {
+      withTempPath { dir =>
+        var pathOne = s"${dir.getCanonicalPath}/table1"
+        (1 to 3).map(i => (i, i.toString)).toDF("a", "b").write.parquet(pathOne)
+        var pathTwo = s"${dir.getCanonicalPath}/table2"
+        (1 to 3).map(i => (i, i.toString)).toDF("c", "b").write.parquet(pathTwo)
+
+        // If the "c = 1" filter gets pushed down, this query will throw an exception which
+        // Parquet emits. This is a Parquet issue (PARQUET-389).
+        checkAnswer(
+          sqlContext.read.parquet(pathOne, pathTwo).filter("c = 1"),
```

Comment:
I investigated that. It does not guarantee the order.

This is because of `FileStatusCache` in `HadoopFsRelation` (which `ParquetRelation` extends as you know). When `FileStatusCache.listLeafFiles()` is called, this returns `Set[FileStatus]` which messes up the order of `Array[FileStatus]`.

So, after retrieving the list of leaf files including `_metadata` and `_common_metadata`,  this starts to merge (separately if necessary) the `Set`s of `_metadata`, `_common_metadata` and part-files in `ParquetRelation.mergeSchemasInParallel()`.

I think this can be resolved by using `LinkedHashSet`.

I will open an issue for this.